File: app/src/main/java/com/gmail/fattazzo/formula1world/ergast/json/service/OnlineDataService.java
package com.gmail.fattazzo.formula1world.ergast.json.service;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.gmail.fattazzo.formula1world.domain.F1Constructor;
import com.gmail.fattazzo.formula1world.domain.F1ConstructorStandings;
import com.gmail.fattazzo.formula1world.domain.F1Driver;
import com.gmail.fattazzo.formula1world.domain.F1DriverStandings;
import com.gmail.fattazzo.formula1world.domain.F1Qualification;
import com.gmail.fattazzo.formula1world.domain.F1Race;
import com.gmail.fattazzo.formula1world.domain.F1Result;
import com.gmail.fattazzo.formula1world.ergast.Ergast;
import com.gmail.fattazzo.formula1world.ergast.json.ErgastManager;
import com.gmail.fattazzo.formula1world.ergast.json.objects.ConstructorStandings;
import com.gmail.fattazzo.formula1world.ergast.json.objects.DriverStandings;
import com.gmail.fattazzo.formula1world.ergast.json.objects.Qualification;
import com.gmail.fattazzo.formula1world.ergast.json.objects.RaceResult;
import com.gmail.fattazzo.formula1world.ergast.json.objects.RaceResults;
import com.gmail.fattazzo.formula1world.ergast.json.objects.Schedule;
import com.gmail.fattazzo.formula1world.service.IDataService;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;

import java.util.ArrayList;
import java.util.List;

/**
 * @author fattazzo
 *         <p/>
 *         date: 11/04/17
 */
@EBean(scope = EBean.Scope.Singleton)
public class OnlineDataService implements IDataService {

    @Bean
    ErgastManager ergastManager;

    @AfterInject
    void afterInjenction() {
        ergastManager.getErgast().setSeason(Ergast.CURRENT_SEASON);
    }

    @Override
    @NonNull
    public List<F1Driver> loadDrivers() {
        List<F1Driver> drivers = new ArrayList<>();
        try {
            List<com.gmail.fattazzo.formula1world.ergast.json.objects.Driver> jsonDrivers = ergastManager.getDrivers();
            for (com.gmail.fattazzo.formula1world.ergast.json.objects.Driver jsonDriver : jsonDrivers) {
                drivers.add(jsonDriver.toF1Driver());
            }
        } catch (Exception e) {
            drivers = new ArrayList<>();
        }
        return drivers;
    }

    /**
     * Load all constructors of the current season.
     *
     * @return constructors
     */
    @Override
    @NonNull
    public List<F1Constructor> loadConstructors() {
        List<F1Constructor> constructors = new ArrayList<>();
        try {
            List<com.gmail.fattazzo.formula1world.ergast.json.objects.Constructor> jsonConstructors = ergastManager.getConstructors();
            for (com.gmail.fattazzo.formula1world.ergast.json.objects.Constructor jsonConstructor : jsonConstructors) {
                constructors.add(jsonConstructor.toF1Constructor());
            }
        } catch (Exception e) {
            constructors = new ArrayList<>();
        }
        return constructors;
    }

    @Override
    @NonNull
    public List<F1Result> loadDriverRacesResult(F1Driver driver) {
        List<F1Result> results = new ArrayList<>();
        try {
            List<RaceResults> jsonResults = ergastManager.getDriverRacersResult(driver.driverRef);
            for (RaceResults jsonresult : jsonResults) {
                results.addAll(jsonresult.toF1Result());
            }
        } catch (Exception e) {
            results = new ArrayList<>();
        }
        return results;
    }

    @Override
    @NonNull
    public List<F1Result> loadConstructorRacesResult(F1Constructor constructor) {
        List<F1Result> results = new ArrayList<>();
        try {
            List<RaceResults> jsonResults = ergastManager.getConstructorRacersResult(constructor.constructorRef);
            for (RaceResults jsonresult : jsonResults) {
                results.addAll(jsonresult.toF1Result());
            }
        } catch (Exception e) {
            results = new ArrayList<>();
        }
        return results;
    }

    @Override
    @NonNull
    public List<F1DriverStandings> loadDriverStandings() {
        List<F1DriverStandings> f1DriverStandings = new ArrayList<>();
        try {
            List<DriverStandings> jsonDriverStandings = ergastManager.getDriverStandings(Ergast.NO_ROUND);
            for (DriverStandings jsonDriverStanding : jsonDriverStandings) {
                f1DriverStandings.add(jsonDriverStanding.toF1DriverStandings());
            }
        } catch (Exception e) {
            f1DriverStandings = new ArrayList<>();
        }

        return f1DriverStandings;
    }

    @Override
    @Nullable
    public F1DriverStandings loadDriverLeader() {

        DriverStandings driverStandings = ergastManager.getDriverLeader();
        if (driverStandings != null) {
            return driverStandings.toF1DriverStandings();
        }
        return null;
    }

    @Override
    @NonNull
    public List<F1ConstructorStandings> loadConstructorStandings() {
        List<F1ConstructorStandings> f1ConstructorStandings = new ArrayList<>();
        try {
            List<ConstructorStandings> jsonConstructorStandings = ergastManager.getConstructorStandings(Ergast.NO_ROUND);
            for (ConstructorStandings jsonConstructorStanding : jsonConstructorStandings) {
                f1ConstructorStandings.add(jsonConstructorStanding.toF1ConstructorStandings());
            }
        } catch (Exception e) {
            f1ConstructorStandings = new ArrayList<>();
        }

        return f1ConstructorStandings;
    }

    @NonNull
    @Override
    public List<F1Race> loadRaces() {
        List<Schedule> schedules;
        try {
            schedules = ergastManager.getSchedules();
        } catch (Exception e) {
            schedules = new ArrayList<>();
        }

        List<F1Race> f1Races = new ArrayList<>();

        for (Schedule schedule : schedules) {
            f1Races.add(schedule.toF1Race());
        }

        return f1Races;
    }

    @NonNull
    @Override
    public List<F1Result> loadRaceResult(F1Race race) {
        List<F1Result> results = new ArrayList<>();
        try {
            List<RaceResult> jsonResults = ergastManager.getRaceResults(race.round);
            for (RaceResult jsonresult : jsonResults) {
                results.add(jsonresult.toF1Result());
            }
        } catch (Exception e) {
            results = new ArrayList<>();
        }
        return results;
    }

    @NonNull
    @Override
    public List<F1Qualification> loadQualification(F1Race race) {
        List<F1Qualification> results = new ArrayList<>();

        try {
            List<Qualification> jsonResults = ergastManager.getQualificationResults(race.round);
            for (Qualification jsonQualification : jsonResults) {
                results.add(jsonQualification.toF1Qualification(race));
            }
        } catch (Exception e) {
            results = new ArrayList<>();
        }

        return results;
    }
}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/ergast/json/service/OnlineDataService.kt
/*
 * Project: total-gp-world
 * File: OnlineDataService.kt
 *
 * Created by fattazzo
 * Copyright © 2018 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.formula1world.ergast.json.service

import com.gmail.fattazzo.formula1world.domain.*
import com.gmail.fattazzo.formula1world.ergast.Ergast
import com.gmail.fattazzo.formula1world.ergast.IDataService
import com.gmail.fattazzo.formula1world.ergast.json.ErgastManager
import com.gmail.fattazzo.formula1world.ergast.json.objects.Schedule
import org.androidannotations.annotations.AfterInject
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.EBean
import java.util.*

/**
 * @author fattazzo
 *
 *
 * date: 11/04/17
 */
@EBean(scope = EBean.Scope.Singleton)
open class OnlineDataService : IDataService {

    @Bean
    lateinit internal var ergastManager: ErgastManager

    @AfterInject
    internal fun afterInjenction() {
        ergastManager!!.ergast.season = Ergast.CURRENT_SEASON
    }

    override fun loadSeason(year: Int): F1Season {
        val f1Season = F1Season()
        f1Season.year = ergastManager!!.ergast.season
        f1Season.current = true
        return f1Season
    }

    override fun updateSeason(season: F1Season) {
        // Nothing to do
    }

    override fun loadDrivers(): List<F1Driver> {
        var drivers: MutableList<F1Driver> = ArrayList()
        try {
            val jsonDrivers = ergastManager!!.drivers
            for (jsonDriver in jsonDrivers) {
                drivers.add(jsonDriver.toF1Driver())
            }
        } catch (e: Exception) {
            drivers = ArrayList()
        }

        return drivers
    }

    /**
     * Load all constructors of the current season.
     *
     * @return constructors
     */
    override fun loadConstructors(): List<F1Constructor> {
        var constructors: MutableList<F1Constructor> = ArrayList()
        try {
            val jsonConstructors = ergastManager!!.constructors
            for (jsonConstructor in jsonConstructors) {
                constructors.add(jsonConstructor.toF1Constructor())
            }
        } catch (e: Exception) {
            constructors = ArrayList()
        }

        return constructors
    }

    override fun loadDriverRacesResult(driver: F1Driver): List<F1Result> {
        var results: MutableList<F1Result> = ArrayList()
        try {
            val jsonResults = ergastManager!!.getDriverRacersResult(driver.driverRef!!)
            for (jsonresult in jsonResults) {
                results.addAll(jsonresult.toF1Result())
            }
        } catch (e: Exception) {
            results = ArrayList()
        }

        return results
    }

    override fun loadConstructorRacesResult(constructor: F1Constructor): List<F1Result> {
        var results: MutableList<F1Result> = ArrayList()
        try {
            val jsonResults = ergastManager!!.getConstructorRacersResult(constructor.constructorRef!!)
            for (jsonresult in jsonResults) {
                results.addAll(jsonresult.toF1Result())
            }
        } catch (e: Exception) {
            results = ArrayList()
        }

        return results
    }

    override fun loadDriverStandings(): List<F1DriverStandings> {
        var f1DriverStandings: MutableList<F1DriverStandings> = ArrayList()
        try {
            val jsonDriverStandings = ergastManager!!.getDriverStandings(Ergast.NO_ROUND)
            for (jsonDriverStanding in jsonDriverStandings) {
                f1DriverStandings.add(jsonDriverStanding.toF1DriverStandings())
            }
        } catch (e: Exception) {
            f1DriverStandings = ArrayList()
        }

        return f1DriverStandings
    }

    override fun loadDriverLeader(): F1DriverStandings? {

        val driverStandings = ergastManager!!.driverLeader
        return driverStandings?.toF1DriverStandings()
    }

    override fun loadConstructorStandings(): List<F1ConstructorStandings> {
        var f1ConstructorStandings: MutableList<F1ConstructorStandings> = ArrayList()
        try {
            val jsonConstructorStandings = ergastManager!!.getConstructorStandings(Ergast.NO_ROUND)
            for (jsonConstructorStanding in jsonConstructorStandings) {
                f1ConstructorStandings.add(jsonConstructorStanding.toF1ConstructorStandings())
            }
        } catch (e: Exception) {
            f1ConstructorStandings = ArrayList()
        }

        return f1ConstructorStandings
    }

    override fun loadRaces(): List<F1Race> {
        var schedules: List<Schedule>
        try {
            schedules = ergastManager!!.schedules
        } catch (e: Exception) {
            schedules = ArrayList()
        }

        val f1Races = ArrayList<F1Race>()

        for (schedule in schedules) {
            f1Races.add(schedule.toF1Race())
        }

        return f1Races
    }

    override fun loadRaceResult(race: F1Race): List<F1Result> {
        var results: MutableList<F1Result> = ArrayList()
        try {
            val jsonResults = ergastManager!!.getRaceResults(race.round)
            for (jsonresult in jsonResults) {
                results.add(jsonresult.toF1Result())
            }
        } catch (e: Exception) {
            results = ArrayList()
        }

        return results
    }

    override fun loadQualification(race: F1Race): List<F1Qualification> {
        var results: MutableList<F1Qualification> = ArrayList()

        try {
            val jsonResults = ergastManager!!.getQualificationResults(race.round)
            for (jsonQualification in jsonResults) {
                results.add(jsonQualification.toF1Qualification(race))
            }
        } catch (e: Exception) {
            results = ArrayList()
        }

        return results
    }

    override fun loadPitStops(race: F1Race): List<F1PitStop> {
        var results: MutableList<F1PitStop> = ArrayList()

        val drivers = loadDrivers()

        try {
            val jsonResults = ergastManager!!.getRacePitStops(race.round)
            for (jsonPit in jsonResults) {
                results.addAll(jsonPit.toF1PitStop(race, drivers))
            }
        } catch (e: Exception) {
            results = ArrayList()
        }

        return results
    }

    override fun loadLaps(race: F1Race, driver: F1Driver): List<F1LapTime> {
        var results: MutableList<F1LapTime> = ArrayList()

        val oldLimit = ergastManager!!.ergast.getLimit()
        try {
            // increase limit for load all race laptimes
            ergastManager!!.ergast.setLimit(150)
            val jsonLaps = ergastManager!!.getLapTimes(race.round, driver.driverRef)
            for (jsonLaps1 in jsonLaps) {
                for (lap in jsonLaps1.getLaps()) {
                    val f1LapTime = lap.toF1LapTime(race, driver)
                    results.add(f1LapTime)
                }
            }
        } catch (e: Exception) {
            results = ArrayList()
        } finally {
            // restore old limit
            ergastManager!!.ergast.setLimit(oldLimit)
        }
        return results
    }
}
